### ISSUE TITLE
Adding XML inside Extensions

### ISSUE THREAD

**Issue by jstoppa** [ITfoxtec/ITfoxtec.Identity.Saml2#96]
I'm trying to create a Saml2AuthnRequest using PostBinding, by following the examples I was able to create the request.

My only problem is that the IdP I'm working with requires me to send a piece of XML inside the Extensions node and when I put the  XML as string, the text ends up being encoded 

For instance, if I create the Extensions like this...
```c#
var extensions = new Extensions()
{
    Data = "<TSP-ID>-</TSP-ID><TSP-TYPE>-</TSP-TYPE>"
}
```
...the Extensions node ends up being set like the below

```xml
<saml2p:Extensions xmlns="urn:oasis:names:tc:SAML:2.0:protocol">&lt;TSP-ID&gt;-&lt;/TSP-ID&gt;&lt;TSP-TYPE&gt;-&lt;/TSP-TYPE&gt;</saml2p:Extensions>
```
which is not what the IdP expects

I looked at the code in Extensions and I can see the XElement just adds the string ([see here](https://github.com/ITfoxtec/ITfoxtec.Identity.Saml2/blob/953ebb2f04f5cb111875bcd7b020c21b80182778/src/ITfoxtec.Identity.Saml2/Schemas/Extensions.cs#L28))

Is there anyway to add XML in Extensions or this requires a code change in the library?

Thanks

**Comment by nixxcz** [comment]
I'm facing the same problem.

I need to compose request like this, which is not possible with ITfoxtec.Identity.Saml2 right now.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<saml2p:AuthnRequest xmlns:saml2p="urn:oasis:names:tc:SAML:2.0:protocol"
    xmlns:saml2="urn:oasis:names:tc:SAML:2.0:assertion" ID="id19cd34deb3c140de8c6eb6790da3de13" Version="2.0" IssueInstant="2018-03-26T14:31:54Z" Destination="https://example.com/abc" AssertionConsumerServiceURL="https://example.com/abc/">
    <saml2:Issuer>https://example.com/abc/</saml2:Issuer>
    <saml2p:RequestedAuthnContext Comparison="minimum">
        <saml2:AuthnContextClassRef>http://eidas.europa.eu/LoA/low</saml2:AuthnContextClassRef>
    </saml2p:RequestedAuthnContext>
    <saml2p:Extensions xmlns:saml2p="urn:oasis:names:tc:SAML:2.0:protocol"
        xmlns:eidas="http://eidas.europa.eu/saml-extensions">
        <eidas:SPType>public</eidas:SPType>
        <eidas:RequestedAttributes>
            <eidas:RequestedAttribute Name="http://eidas.europa.eu/attributes/naturalperson/PersonIdentifier" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
            <eidas:RequestedAttribute Name="http://eidas.europa.eu/attributes/naturalperson/CurrentGivenName" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
            <eidas:RequestedAttribute Name="http://eidas.europa.eu/attributes/naturalperson/CurrentFamilyName" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
            <eidas:RequestedAttribute Name="http://eidas.europa.eu/attributes/naturalperson/CurrentAddress" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
            <eidas:RequestedAttribute Name="http://eidas.europa.eu/attributes/naturalperson/DateOfBirth" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
            <eidas:RequestedAttribute Name="http://eidas.europa.eu/attributes/naturalperson/PlaceOfBirth" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
            <eidas:RequestedAttribute Name="http://www.stork.gov.eu/1.0/countryCodeOfBirth" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
            <eidas:RequestedAttribute Name="http://www.stork.gov.eu/1.0/eMail" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
            <eidas:RequestedAttribute Name="http://www.stork.gov.eu/1.0/age" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
            <eidas:RequestedAttribute Name="http://www.stork.gov.eu/1.0/isAgeOver" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false">
                <eidas:AttributeValue>18</eidas:AttributeValue>
            </eidas:RequestedAttribute>
            <eidas:RequestedAttribute Name="http://schemas.eidentity.cz/moris/2016/identity/claims/phonenumber" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
            <eidas:RequestedAttribute Name="http://schemas.eidentita.cz/moris/2016/identity/claims/tradresaid" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
            <eidas:RequestedAttribute Name="http://schemas.eidentita.cz/moris/2016/identity/claims/idtype" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
            <eidas:RequestedAttribute Name="http://schemas.eidentita.cz/moris/2016/identity/claims/idnumber" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
        </eidas:RequestedAttributes>
    </saml2p:Extensions>
</saml2p:AuthnRequest>```

**Comment by Revsgaard** [comment]
I have implemented a solution, please take a look at the https://github.com/ITfoxtec/ITfoxtec.Identity.Saml2/commit/d929c88f77a19702e9c914c7a25ffaf96ffe44f1 commit and let me know if it works for you.

The requested XML is added to the TestWebAppCore sample in the AppExtensions.cs.
The TestWebAppCore sample should generate the following XML if you uncomment the `Extensions = new AppExtensions(),` line in AuthController.cs:

```xml
<saml2p:AuthnRequest xmlns:saml2p="urn:oasis:names:tc:SAML:2.0:protocol" xmlns:saml2="urn:oasis:names:tc:SAML:2.0:assertion" ID="_6474211f-b7b0-46d4-a0a6-ff256c917b95" Version="2.0" IssueInstant="2021-11-02T20:13:26.447Z" Destination="https://localhost:44305/Auth/Login">
  <saml2:Issuer>itfoxtec-testwebappcore</saml2:Issuer>
  <saml2p:Extensions xmlns:eidas="http://eidas.europa.eu/saml-extensions">
    <eidas:SPType>public</eidas:SPType>
    <eidas:RequestedAttributes>
      <eidas:RequestedAttribute Name="http://eidas.europa.eu/attributes/naturalperson/PersonIdentifier" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
      <eidas:RequestedAttribute Name="http://eidas.europa.eu/attributes/naturalperson/CurrentGivenName" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
      <eidas:RequestedAttribute Name="http://eidas.europa.eu/attributes/naturalperson/CurrentFamilyName" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
      <eidas:RequestedAttribute Name="http://eidas.europa.eu/attributes/naturalperson/CurrentAddress" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
      <eidas:RequestedAttribute Name="http://eidas.europa.eu/attributes/naturalperson/DateOfBirth" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
      <eidas:RequestedAttribute Name="http://eidas.europa.eu/attributes/naturalperson/PlaceOfBirth" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
      <eidas:RequestedAttribute Name="http://www.stork.gov.eu/1.0/countryCodeOfBirth" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
      <eidas:RequestedAttribute Name="http://www.stork.gov.eu/1.0/eMail" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
      <eidas:RequestedAttribute Name="http://www.stork.gov.eu/1.0/age" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
      <eidas:RequestedAttribute Name="http://www.stork.gov.eu/1.0/isAgeOver" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false">
        <eidas:AttributeValue>18</eidas:AttributeValue>
      </eidas:RequestedAttribute>
      <eidas:RequestedAttribute Name="http://schemas.eidentity.cz/moris/2016/identity/claims/phonenumber" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
      <eidas:RequestedAttribute Name="http://schemas.eidentity.cz/moris/2016/identity/claims/tradresaid" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
      <eidas:RequestedAttribute Name="http://schemas.eidentity.cz/moris/2016/identity/claims/idtype" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
      <eidas:RequestedAttribute Name="http://schemas.eidentity.cz/moris/2016/identity/claims/idnumber" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri" isRequired="false" />
    </eidas:RequestedAttributes>
  </saml2p:Extensions>
  <saml2:Subject>
    <saml2:NameID>abcd</saml2:NameID>
  </saml2:Subject>
  <saml2p:NameIDPolicy AllowCreate="true" Format="urn:oasis:names:tc:SAML:2.0:nameid-format:persistent" />
</saml2p:AuthnRequest>
```

**Comment by jstoppa** [comment]
@Revsgaard thanks for this and sorry I didn't get back to you before 

I'll give it a try this week

**Comment by nixxcz** [comment]
@Revsgaard Thank you! I'm just looking into it and it looks good! Could you publish it as a (perhaps pre-release) NuGet so I don't have to include the source in my project?

**Comment by nixxcz** [comment]
It seems to work great, thank you! 😃 👍

**Comment by Revsgaard** [comment]
Thanks for getting back. I have released the support in a beta https://github.com/ITfoxtec/ITfoxtec.Identity.Saml2/releases/tag/4.7.1-beta1

**Comment by nixxcz** [comment]
Awesome, I have upgraded my solution via NuGet and it works flawlessly so far. 🙂
Just beware - you probably published the NuGet as 4.7.1 **stable** (without the _pre-release_ tag), regardless of the GitHub release tag.

**Comment by Revsgaard** [comment]
Thank you for getting back :)
It was an error to publish the NuGet 4.7.1 as stable. But locally it works so I have upgraded det release to stable in GitHub as well 
https://github.com/ITfoxtec/ITfoxtec.Identity.Saml2/releases/tag/4.7.1
